pkg/providers/tinkerbell/create.go

```diff
@@ -45,7 +45,15 @@ func (p *Provider) PreCAPIInstallOnBootstrap(ctx context.Context, cluster *types
 }
 
 func (p *Provider) PostBootstrapSetup(ctx context.Context, clusterConfig *v1alpha1.Cluster, cluster *types.Cluster) error {
-	return p.applyHardware(ctx, cluster)
+	if err := p.applyHardware(ctx, cluster); err != nil {
+		return err
+	}
+	hardwareSpec, err := p.generateHardwareSpec(ctx, cluster)
```

Comment:
I think `p.applyHardware()` also generates the hardware spec since it has to apply it. Can we consolidate these two calls somehow so we are not generating the hardware spec twice?

Reply:
The spec that is generated for `p.applyHardware()` is partial and deals with the hardware fed from CSV file (or in general added hardware in the case of upgrade I think). The hardware yaml we want to generate must include all of the hardware in the management cluster  and the related objects. We probably don't want `p.applyHardware` to try to apply existing hardware, hence this approach. So I'm not sure if we can merge them. and

Reply:
As a `PostBootstrapSetup` process, we apply hardwares by marshaling the catalogue object. Could you elaborate a little about what is the main purpose of generateHardwareSpec and buildHardwareMachineFromCluster functions?